bftengine/tests/testViewChange/test_views_manager.cpp

```diff
@@ -0,0 +1,208 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#include "gtest/gtest.h"
+#include "helper.hpp"
+#include "Logger.hpp"
+#include "ReplicaConfig.hpp"
+#include "SigManager.hpp"
+#include "ViewsManager.hpp"
+
+namespace {
+
+using namespace bftEngine;
+using namespace bftEngine::impl;
+
+constexpr int numberOfFaultyReplicas = 2;
+constexpr bool dynamicCollectorForPartialProofs = true, dynamicCollectorForExecutionProofs = false;
+constexpr int initialView = 0;
+constexpr bftEngine::impl::SeqNum lastStableSeqNum = 150;
+
+class ViewsManagerTest : public ::testing::Test {
+ protected:
+  ReplicaConfig& rc;
+  ReplicasInfo replicasInfo;
+  std::unique_ptr<SigManager> sigManager;
+  std::unique_ptr<ViewsManager> viewsManager;
+
+  ViewsManagerTest()
+      : rc(createReplicaConfig(numberOfFaultyReplicas)),
+        replicasInfo(rc, dynamicCollectorForPartialProofs, dynamicCollectorForExecutionProofs),
+        sigManager(createSigManager(rc.replicaId,
+                                    rc.replicaPrivateKey,
+                                    KeyFormat::HexaDecimalStrippedFormat,
+                                    rc.publicKeysOfReplicas,
+                                    replicasInfo)),
+        viewsManager(std::make_unique<ViewsManager>(&replicasInfo)) {}
+};
+
+TEST_F(ViewsManagerTest, moving_to_higher_view) {
+  const auto currentView = viewsManager->getCurrentView();
+  LOG_INFO(GL, KVLOG(currentView));
+  viewsManager->setHigherView(currentView + 1);
+  LOG_INFO(GL, KVLOG(viewsManager->getCurrentView()));
+  ASSERT_EQ(viewsManager->getCurrentView(), currentView + 1);
+}
+
+TEST_F(ViewsManagerTest, store_complaint) {
+  std::unique_ptr<ReplicaAsksToLeaveViewMsg> complaint(ReplicaAsksToLeaveViewMsg::create(
+      rc.replicaId, initialView, ReplicaAsksToLeaveViewMsg::Reason::ClientRequestTimeout));
+
+  viewsManager->storeComplaint(std::move(complaint));
```

Comment:
I see that here we store the Complaint unconditionally. Might be it will be better to transfer the check whether we need this complaint from [ReplicaImp.cpp:301](https://github.com/vmware/concord-bft/blob/d3cd2e781598f6f50cca35420d32263eb96c11ad/bftengine/src/bftengine/ReplicaImp.cpp#L301) to this method.

Reply:
as discussed with @StefanPalashev and @HristoStaykov , this will be addressed in a separate PR.